### ISSUE TITLE
Error: Could not add AttachedObjectHandler to MyFaces, after migration to Websphere Liberty 8.5.5?

### ISSUE THREAD

**Issue by NottyCode** [primefaces/primefaces#680]
Issue 7998 from https://code.google.com/p/primefaces/issues/detail?id=7998 appears to affect WAS Liberty 8.5.5.6 running JSF-2.2 which is based on MyFaces 2.2.8 as reported on stack overflow here: http://stackoverflow.com/questions/32370745/error-could-not-add-attachedobjecthandler-to-myfaces-after-migration-to-websph While the original issue was closed in Feb because 2.0.5 of myfaces is old.

Specific failure is:

[ERROR   ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext

**Comment by tandraschko** [comment]
Please post the complete stacktrace.

**Comment by anatoly314** [comment]
It's same message that returns to itself, i.e. :

[ERROR ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext
[ERROR ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext
[ERROR ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext

**Comment by anatoly314** [comment]
It's the same message that returns to itself, I can attach a screenshot if it may help you:

[ERROR   ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext
[ERROR   ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext
[ERROR   ] Could not add AttachedObjectHandler to MyFaces!
org.apache.myfaces.view.facelets.FaceletCompositionContext

**Comment by tandraschko** [comment]
I need the complete stacktrace, not just the first line.

**Comment by tandraschko** [comment]
JFYI: it works fine with MyFaces 2.2.8. So no idea whats wrong on WAS.
Maybe a WAS issue or classloading issue. Try talk with the WAS guys and come back.

**Comment by tandraschko** [comment]
Just some infos from my side:
There is no way in the JSF API to add an "AttachedObjectHandler" if the client behavior is used on a composite component.
In PrimeFaces we need to do this via a hack for both Mojarra and MyFaces. See AbstractBehaviorHandler line 218-240

**Comment by anatoly314** [comment]
WAS guy has opened this issue

**Comment by tandraschko** [comment]
I see.

@notatibm Please come back if you need more infos or if you have an idea how to fix this on our side. But AFAICS we can't do anything - our hack works fine on MyFaces 2.2.8.

**Comment by NottyCode** [comment]
So reading the code in AbstractBehaviorHandler  line 218-240 the MyFaces workaround wont work by default because the MyFaces classes are not exposed to the applicaation classloader.

**Comment by NottyCode** [comment]
If WAS Liberty were to be updated such that org.apache.myfaces.view.facelets.FaceletCompositionContext were available to the thread context classloader would prime faces consider updating AbstractBehaviorHandler to use the thread context classloader in place of the classloader used to load AbstractBehaviorHandler? Note in WAS Liberty these are not the same thing, the thread context classloader sees a super set of classes.

**Comment by tandraschko** [comment]
Thats what i thouht but i dont see any other possible solution.

**Comment by tandraschko** [comment]
Would be ok from my side. 
Would you please try that on your side an come back with a patch?

**Comment by tandraschko** [comment]
any news?

**Comment by tandraschko** [comment]
I will do a fallback to the ThreadClassloader for the next release.

**Comment by tandraschko** [comment]
added the fallback - please create a new issue if we can still enhance the WAS compatibility.

**Comment by lostiniceland** [comment]
Liberty is OSGi based and only exports javax.faces-packages to the JEE-world, but org.primefaces.behavior.base.AbstractBehaviorHandler is trying to initialize a Class during runtime

```
Class.forName("org.apache.myfaces.view.facelets.FaceletCompositionContext")
```

Since the package is not exported you get a ClassNotFoundException.

Here is a rather dirty hack.
- open libertyInstallFolder/lib/com.ibm.ws.jsf_1.0.10.jar
- extract MANIFEST.MF and edit it
  - add org.apache.myfaces.view.facelets;version="2.0.2" to the Export-Package-header (do not exceed  the line-length)
- put the modified MANIFEST.MF back into the archive
- open libertyInstallFolder/lib/features/jsf-2.0.mf
  - add org.apache.myfaces.view.facelets to the IBM-API-header (this will bridge from OSGi to the JEE world)
- start Liberty profile

This works for JSF 2.0, for 2.2 you have to use the com.ibm.ws.jsf.2.2_1.0.10.jar and jsf-2.2.mf.

I will report this to the Liberty-devs, but I am not sure if they are willing to export implementation details, so it might be best if IBM and Primefaces would work together on this issue.

**Comment by tandraschko** [comment]
I already added a fallback to the TCCL, we can't do anything else on PF side.

**Comment by tandraschko** [comment]
I also try to reanimate the spec issue here: https://java.net/jira/browse/JAVASERVERFACES_SPEC_PUBLIC-922
You could ping the spec mailing list that this issue is very important for some EE containers which uses MyFaces.

**Comment by NottyCode** [comment]
We are shipping our part of the fix for this with the WAS Liberty 8.5.5.8 fix. It is documented here: http://www-01.ibm.com/support/docview.wss?uid=swg1PI47095 You'll need the primefaces with this fix as well.
